crates/re_query_cache2/src/range/results.rs

```diff
@@ -0,0 +1,747 @@
+
+    time_range: TimeRange,
+    front_status: PromiseResult<()>,
+    back_status: PromiseResult<()>,
+
+    /// Keeps track of reentrancy counts for the current thread.
+    ///
+    /// Used to detect and prevent potential deadlocks when using the cached APIs in work-stealing
+    /// environments such as Rayon.
+    reentering: &'static std::thread::LocalKey<RefCell<u32>>,
+}
+
+impl<'a, T> Drop for CachedRangeData<'a, T> {
+    #[inline]
+    fn drop(&mut self) {
+        self.reentering
+            .with_borrow_mut(|reentering| *reentering = reentering.saturating_sub(1));
+    }
+}
+
+impl<'a, T> CachedRangeData<'a, T> {
+    /// Returns the current status on both ends of the range.
+    ///
+    /// E.g. it is possible that the front-side of the range is still waiting for pending data while
+    /// the back-side has been fully loaded.
+    #[inline]
+    pub fn status(&self) -> (PromiseResult<()>, PromiseResult<()>) {
+        (self.front_status.clone(), self.back_status.clone())
+    }
+
+    #[inline]
+    pub fn range_indices(
+        &self,
+        entry_range: Range<usize>,
+    ) -> impl Iterator<Item = &(TimeInt, RowId)> {
+        match self.indices.as_ref() {
+            Some(indices) => itertools::Either::Left(indices.range(entry_range)),
+            None => itertools::Either::Right(std::iter::empty()),
+        }
+    }
+
+    #[inline]
+    pub fn range_data(&self, entry_range: Range<usize>) -> impl Iterator<Item = &[T]> {
+        match self.data.as_ref() {
+            Some(indices) => itertools::Either::Left(indices.range(entry_range)),
+            None => itertools::Either::Right(std::iter::empty()),
+        }
+    }
+
+    /// Range both the indices and data by zipping them together.
+    ///
+    /// Useful for time-based joins (`range_zip`).
+    #[inline]
+    pub fn range_indexed(&self) -> impl Iterator<Item = (&(TimeInt, RowId), &[T])> {
+        let entry_range = self.entry_range();
+        itertools::izip!(
+            self.range_indices(entry_range.clone()),
+            self.range_data(entry_range)
+        )
+    }
+
+    /// Returns the index range that corresponds to the specified `time_range`.
+    ///
+    /// Use the returned range with one of the range iteration methods:
+    /// - [`Self::range_indices`]
+    /// - [`Self::range_data`]
+    /// - [`Self::range_indexed`]
+    ///
+    /// Make sure that the bucket hasn't been modified in-between!
+    ///
+    /// This is `O(2*log(n))`, so make sure to clone the returned range rather than calling this
+    /// multiple times.
+    #[inline]
+    pub fn entry_range(&self) -> Range<usize> {
+        let Some(indices) = self.indices.as_ref() else {
+            return 0..0;
+        };
+
+        // If there's any static data cached, make sure to look for it explicitly.
+        //
+        // Remember: `TimeRange`s can never contain `TimeInt::STATIC`.
+        let static_override = if matches!(indices.front(), Some((TimeInt::STATIC, _))) {
```

Comment:
nit/style

```suggestion
        let static_override = if indices.front() == Some((TimeInt::STATIC, _)) {
```

Reply:
i would prefer this but `_` is not valid as RHS outside of matches